sklearn/metrics/cluster/_dbcv_internals.py

```diff
@@ -0,0 +1,301 @@
+
+    References
+    ----------
+    Moulavi, D., Jaskowiak, P.A., Campello, R.J., Zimek, A. and Sander, J.,
+    2014. Density-Based Clustering Validation. In SDM (pp. 839-847).
+    """
+    if metric == "precomputed":
+        sub_select = X[labels == cluster_id1, :][:, labels == cluster_id2]
+        distance_matrix = sub_select[internal_nodes1, :][:, internal_nodes2]
+    else:
+        cluster1 = X[labels == cluster_id1][internal_nodes1]
+        cluster2 = X[labels == cluster_id2][internal_nodes2]
+        distance_matrix = pairwise_distances(cluster1, cluster2, metric, **kwd_args)
+
+    if no_coredist:
+        return distance_matrix.min()
+
+    else:
+        core_dist_matrix1 = np.tile(
+            core_distances1[internal_nodes1], (distance_matrix.shape[1], 1)
+        ).T
+        core_dist_matrix2 = np.tile(
+            core_distances2[internal_nodes2], (distance_matrix.shape[0], 1)
+        )
+
+        mr_dist_matrix = np.dstack(
+            [distance_matrix, core_dist_matrix1, core_dist_matrix2]
+        ).max(axis=-1)
+
+        return mr_dist_matrix.min()
+
+
+def internal_minimum_spanning_tree(mr_distances):
+    """
+    Compute the 'internal' minimum spanning tree given a matrix of mutual
+    reachability distances. Given a minimum spanning tree the 'internal'
+    graph is the subgraph induced by vertices of degree greater than one.
+
+    Parameters
+    ----------
+    mr_distances : array (cluster_size, cluster_size)
+        The pairwise mutual reachability distances, inferred to be the edge
+        weights of a complete graph. Since MSTs are computed per cluster
+        this is the all-points-mutual-reachability for points within a single
+        cluster.
+
+    Returns
+    -------
+    internal_nodes : array
+        An array listing the indices of the internal nodes of the MST
+
+    internal_edges : array (?, 3)
+        An array of internal edges in weighted edge list format; that is
+        an edge is an array of length three listing the two vertices
+        forming the edge and weight of the edge.
+
+    References
+    ----------
+    Moulavi, D., Jaskowiak, P.A., Campello, R.J., Zimek, A. and Sander, J.,
+    2014. Density-Based Clustering Validation. In SDM (pp. 839-847).
+    """
+    from ...cluster._hierarchical_fast import mst_linkage_core
+
+    min_span_tree = mst_linkage_core(mr_distances, None, precomputed=True)
+    # mark internal nodes via converting the occurence count to bool
+    internal_node_flags = np.arange(mr_distances.shape[0])[
+        np.bincount(min_span_tree.T[:2].flatten().astype(np.intp)) > 1
+    ]
+    if not len(internal_node_flags):
+        internal_node_flags = [0]
+    # A little "fancy" we select from the flattened array reshape back
+    # (Fortran format to get indexing right) and take the product to do an and
+    # then convert back to boolean type.
+    edge_selection = np.prod(
+        np.in1d(min_span_tree.T[:2], internal_node_flags).reshape(
```

Comment:
I wasn't familiar with `isin` before this. I had a look at the [source](https://github.com/numpy/numpy/blob/v1.26.0/numpy/lib/arraysetops.py#L766-L891) and I don't feel good about this suggestion. As you probably know, it's just a wrapper around `in1d` that subsequently passes in the original shape into a `reshape` call. That feels redundant/unwieldy since we perform a reshape with custom arguments right after.

Regardless of performance implications I can't say I'm 100% confident the internal reshape wouldn't actually mess with the Fortran style indexing we apply right after? You know more about numpy than me, if you still want to go forwards after taking my concern into consideration I'll let you overrule me on this

Reply:
97023811d2185277a3d705627c2f9e462c0f1447